src/units/abilities.cpp

```diff
@@ -966,14 +989,232 @@ void attack_type::modified_attacks(bool is_backstab, unsigned & min_attacks,
+		const std::string& apply_to = (*i->first)["apply_to"];
+		if(apply_to == "both" || apply_to == "opponent") {
+			return true;
+		}
+		if(attacker && apply_to == "defender") {
+			return true;
+		}
+		if(!attacker && apply_to == "attacker") {
+			return true;
+		}
+		++i;
+	}
+	return false;
+}
+
+unit_ability_list list_leadership(const std::string& ability, unit_const_ptr un, unit_const_ptr up, const map_location& loc, const map_location& opp_loc, bool attacker, const_attack_ptr weapon, const_attack_ptr opp_weapon)
+{
+	unit_ability_list abiln (loc);
+	unit_ability_list abil = (*un).get_abilities(ability, loc);
+	for(unit_ability_list::iterator i = abil.begin(); i != abil.end();) {
+		const config &filter = (*i->first).child("filter_opponent");
+		const config &filter_attacker = (*i->first).child("filter_attacker");
+		const config &filter_defender = (*i->first).child("filter_defender");
+		bool fighter_filter = ability_apply_filter(un, up, ability, *i->first, loc, opp_loc, attacker, weapon, opp_weapon);
+		const std::string& active_on = (*i->first)["active_on"];
+		bool active_on_bool = !(active_on.empty() || (attacker && active_on == "offense") || (!attacker && active_on == "defense"));
+		bool weapon_filter = !((*un).ability_affects_weapon(*i->first, weapon, false) && (*un).ability_affects_weapon(*i->first, opp_weapon, true));
+		if(!up &&(filter || (filter_attacker && !attacker) || (filter_defender && attacker))) {
+			return abiln;
+		}
+
+		if(active_on_bool || fighter_filter || weapon_filter) {
+			i = abil.erase(i);
+		} else {
+			++i;
+		}
+	}
+	return abil;
+}
+
+unit_ability_list attack_type::list_ability(const std::string& ability) const
+{
+	const unit_map& units = display::get_singleton()->get_units();
+
+	unit_const_ptr self = self_;
+	unit_const_ptr other = other_;
+
+	if(self == nullptr) {
+		unit_map::const_iterator it = units.find(self_loc_);
+		if(it.valid()) {
+			self = it.get_shared_ptr().get();
+		}
+	}
+	if(other == nullptr) {
+		unit_map::const_iterator it = units.find(other_loc_);
+		if(it.valid()) {
+			other = it.get_shared_ptr().get();
+		}
+	}
+
+	// Make sure they're facing each other.
+	temporary_facing self_facing(self, self_loc_.get_relative_dir(other_loc_));
+	temporary_facing other_facing(other, other_loc_.get_relative_dir(self_loc_));
+	unit_ability_list abil_list(self_loc_);
+	if(self){
+		if(leadership_affects_self(ability, *self, self_loc_, is_attacker_)) {
+			abil_list = list_leadership(ability, self, other, self_loc_, other_loc_, is_attacker_, shared_from_this(), other_attack_);
+		}
+	}
+	if(other){
+		if(leadership_affects_opponent(ability, *other, other_loc_, !is_attacker_)) {
+			abil_list = list_leadership(ability, other, self, other_loc_, self_loc_, !is_attacker_, other_attack_, shared_from_this());
+		}
+	}
+	if(!abil_list.empty()){	
+		for(const config& i : specials_.child_range(ability)) {
+			if(!get_specials(ability).empty()) {
```

Comment:
this code also doesnt make any sense, you are just checking whether tha unit has any ability with the id/tagname ability instead of checkign whether the ability described by`i` is really active